### PR TITLE
feat(examples): add agent behavior presentation grader

### PR DESCRIPTION
**Pull request by ZhaoAndy821**
## Summary

Adds a new example, `eval-agent-behavior-review`: a deterministic Python grader that scores how an agent's behavior is presented in a session trace. Things like: intent declared before actions, retry convergence, conclusions not buried in reasoning, pacing, vocabulary stability.

Promptfoo's built-in assertions cover correctness, similarity, and LLM judges. This targets a different axis. An agent can complete a task correctly and still behave unreadably. The 12 checks map from Disney's 12 animation principles, which turn out to be a decent rubric for legible agent behavior.

## What's included

- `behavior_review.py` - the grader (12 rules, fail-closed on missing `ok` annotations, returns a `GradingResult` with per-dimension `componentResults` and `namedScores`)
- `promptfooconfig.yaml` - echo provider, no API key needed, 2 demo cases
- `test_behavior_review.py` - 26 unit tests (at least 3 boundary cases per rule)
- `README.md` - principle-to-rule mapping table and usage
- `.gitignore`

## Validation

- `python -m unittest` passes locally (26 tests, Python 3.9+ compatible)
- `promptfoo eval` runs cleanly with the `echo` provider: the `stiff` demo session fails at 0.58, the `improved` demo session passes at 1.00
- `ruff check` and `ruff format --check` clean

### CODE REVIEW

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+            "pass": False,
+            "score": 0.0,
+            "reason": "data_quality low: actions missing ok annotation; retry loop undeterminable",
+        }
+    else:
+        err_retries: dict[str, int] = {}
+        for s in fails:
+            key = (
+                str(s.get("tool")) + "|" + str(s.get("text", ""))[:60]
+            )  # (tool, text) error identity
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Fixed. The retry key now uses the full error text, so failures that share a long prefix are no longer merged into a fake loop. There's a test for that too.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    import json
+
+    raw = (context.get("vars") or {}).get("session")
+    if not raw:
+        return []
+    if isinstance(raw, (list, dict)):
+        steps = raw.get("steps") if isinstance(raw, dict) else raw
+    else:
+        try:
+            steps = json.loads(raw).get("steps")
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Fixed. The loader type-checks the parsed JSON, so a serialized list is graded the same way as a dict with a "steps" key. New test included.

File: examples/eval-agent-behavior-review/promptfooconfig.yaml [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# Agent Behavior Review — promptfoo demo config
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/promptfooconfig.yaml]: Added the schema line and description in the order examples/AGENTS.md wants. Also ran Prettier on the config and README, which should clear the Style Check failure.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+        ante = {"pass": True, "score": 1.0, "reason": "no tool actions to anticipate"}
+    elif first_act == 0:
+        ante = {
+            "pass": False,
+            "score": 0.0,
+            "reason": "first action without prior intent declaration",
+        }
+    else:
+        prev = steps[first_act - 1]
+        ok_ = prev.get("kind") in _PLAN_KINDS
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Right, the strict adjacency check was too brittle. Anticipation now scans the whole prefix before the first tool action, so an explanatory message in between doesn't erase a valid plan. Test added.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+    # 11 SolidDrawing: heuristic (documented) — vocabulary stability
+    import re
+
+    terms = re.findall(r"[A-Za-z][A-Za-z0-9_-]{3,}", text)
+    freq: dict[str, int] = {}
+    for t in terms:
+        freq[t] = freq.get(t, 0) + 1
+    repeated = [t for t, c in freq.items() if c >= 3]
+    ok_ = len(repeated) <= T["max_repeated_terms"]
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py:289]: Fair point about the wording. The metric is deliberately a repeated-term cap (documented heuristic), not a drift detector. I fixed the reason string, docstring and README to say what it actually measures, rather than redefining the dimension in an example.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    low = text.lower()
+    return sum(low.count(k) for k in kws)
+
+
+def _components(steps: list[dict[str, Any]]) -> dict[str, Any]:
+    if not steps:
+        return {
+            "comps": [{"pass": False, "score": 0.0, "reason": "empty session trace"}],
+            "quality": "low",
+        }
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Fixed. _components now returns no components for an empty trace, so the missing-trace branch in get_assert actually runs and the result shape matches the other failure paths.

[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Fixed. _components now returns no components for an empty trace, so the missing-trace branch in get_assert actually runs and the result shape matches the other failure paths.

File: examples/eval-agent-behavior-review/test_behavior_review.py [hunk 7]
```diff
@@ -0,0 +1,358 @@
+"""Unit tests for behavior_review — unittest style (promptfoo contribution guide).
+
+Run: python -m unittest test_behavior_review
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/test_behavior_review.py:3]: True, but wiring examples/ into CI means touching .github/workflows, which felt out of scope for an example contribution. The suite runs with `python -m unittest test_behavior_review` (31 tests now). If a maintainer wants a CI job for it, happy to add one.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+            "score": 1.0 if ok_ else 0.0,
+            "reason": f"repeated terms={len(repeated)} (heuristic)"
+            if ok_
+            else "vocabulary drift suspected (heuristic)",
+        }
+    )
+
+    # 12 Appeal: consistent closing signature
+    closers = ("done", "完成", "summary", "总结", "next steps", "下一步")
+    appeal = any(c in text.lower() for c in closers)
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: The closing check now looks only at the final step, and I added a test that a mid-trace "done" no longer rescues an abrupt ending.

**Comment by ZhaoAndy821** [comment]
Both red jobs look transient rather than diff-related: the redteam integration run hit a 429, and the code scan failed to open a websocket to api.promptfoo.app. Everything touching the example itself is green. A re-run whenever someone gets a chance should clear them.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+            "score": 0.0,
+            "reason": "data_quality low: actions missing ok annotation; retry loop undeterminable",
+        }
+    else:
+        err_retries: dict[str, int] = {}
+        for s in fails:
+            key = (
+                str(s.get("tool")) + "|" + str(s.get("text", ""))
+            )  # full error identity
+            err_retries[key] = err_retries.get(key, 0) + 1
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Converged retries were indeed getting counted as loops. Rule 3 now tracks consecutive failures per error identity and clears the streak on success. `fail-success-fail-success-fail-success` passes now, with a test for it.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+    """count total occurrences of all keywords"""
+    low = text.lower()
+    return sum(low.count(k) for k in kws)
+
+
+def _components(steps: list[dict[str, Any]]) -> dict[str, Any]:
+    if not steps:
+        return {"comps": [], "quality": "low"}
+
+    actions = [s for s in steps if s.get("kind") == "action"]
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py:92]: Fixed. The loader checks that every step is a mapping before grading, so a trace like `{"steps": [null]}` goes through the documented fail-closed path instead of raising.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+        }
+    )
+
+    # 11 SolidDrawing: heuristic (documented) — cap on repeated English terms
+    import re
+
+    terms = re.findall(r"[A-Za-z][A-Za-z0-9_-]{3,}", text)
+    freq: dict[str, int] = {}
+    for t in terms:
+        freq[t] = freq.get(t, 0) + 1
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Good catch, `File`/`file`/`FILE` were three separate terms. Counting is case-insensitive now; regression test included.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+            "score": 0.0,
+            "reason": "no session trace in vars.session",
+            "componentResults": [],
+        }
+    score = sum(c["score"] for c in comps) / len(comps)
+    named = {("p" + str(i + 1)): c["score"] for i, c in enumerate(comps)}
+    fails = [c["reason"] for c in comps if not c["pass"]]
+    data_note = " [data_quality=low]" if res["quality"] == "low" else ""
+    return {
+        "pass": score >= THRESHOLDS["pass_score"],
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Agreed, that contradicted the fail-closed claim in the README. Low data quality now overrides the aggregate pass regardless of score. There's a test asserting a 0.92-but-unverifiable trace fails.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+    return sum(low.count(k) for k in kws)
+
+
+def _components(steps: list[dict[str, Any]]) -> dict[str, Any]:
+    if not steps:
+        return {"comps": [], "quality": "low"}
+
+    actions = [s for s in steps if s.get("kind") == "action"]
+    tool_actions = [s for s in actions if s.get("tool")]
+    fails = [s for s in tool_actions if s.get("ok") is False]
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Fixed. Any action whose `ok` is not a boolean counts as unannotated and fails closed, so `"false"`/`0`/`1` can no longer score as successes. Regression test added for the false-negative path.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+            "score": 1.0 if ok_ else 0.0,
+            "reason": f"max same-error retries={worst} <= {T['max_retries']}"
+            if ok_
+            else f"retry loop x{worst}",
+        }
+    out.append(squash)
+
+    # 4 PoseToPose: true mid-session checkpoint (verify must not be the final step)
+    mid_verify = any(k in _VERIFY_KINDS for k in kinds[1:-1])
+    p2p = (not (len(steps) > T["long_session"])) or mid_verify
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Right, a preflight check shouldn't count. A qualifying checkpoint now needs an action on both sides of it, i.e. it must sit between the first and last action of the run.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+            "score": 1.0 if ok_ else 0.0,
+            "reason": f"aux occurrences={aux} <= {T['max_aux']}"
+            if ok_
+            else f"aux noise too high ({aux})",
+        }
+    )
+
+    # 9 Timing: long sessions emit intermediate feedback
+    mid_msgs = any(k in _VERIFY_KINDS or k == "message" for k in kinds[1:-1])
+    timing = (not (len(steps) > T["very_long_session"])) or mid_msgs
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Same issue as rule 4, fixed the same way: feedback only counts if it arrives after the first action and before the final step.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+    low = text.lower()
+    return sum(low.count(k) for k in kws)
+
+
+def _components(steps: list[dict[str, Any]]) -> dict[str, Any]:
+    if not steps:
+        return {"comps": [], "quality": "low"}
+
+    actions = [s for s in steps if s.get("kind") == "action"]
+    tool_actions = [s for s in actions if s.get("tool")]
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Fixed. Retry analysis now covers every action; when the tool field is omitted the step falls back to an `action|<text>` identity, so three identical failures without tool metadata report a retry loop instead of scoring 1.0. Regression test added.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 17]
```diff
@@ -0,0 +1,308 @@
...
+    topics = [s.get("topic") for s in steps if s.get("topic")]
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Classic truthiness bug, fixed. Topics are now kept whenever present (`is not None`), so numeric ids like 0 survive and `0 -> 1 -> 0` counts as a zig-zag. The README note on topic values was updated to match.

File: examples/eval-agent-behavior-review/behavior_review.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Fixed. Retry analysis now covers every action; when the tool field is omitted the step falls back to an `action|<text>` identity, so three identical failures without tool metadata report a retry loop instead of scoring 1.0. Regression test added.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 18]
```diff
@@ -0,0 +1,308 @@
...
+            "pass": declared,
+            "score": 1.0 if declared else 0.0,
+            "reason": "intent declared before first action"
+            if declared
+            else "no intent declaration before first action",
+        }
+    out.append(ante)
+
+    # 2 Staging: session opens with plan/intent
+    opens_ok = bool(kinds) and kinds[0] in _PLAN_KINDS
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Fixed in the loader: every step must carry a string `kind`, otherwise the trace goes through the fail-closed no-trace path instead of raising on set membership. Test included.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 19]
```diff
@@ -0,0 +1,308 @@
...
+        except (ValueError, TypeError):
+            return []
+        if isinstance(parsed, dict):
+            steps = parsed.get("steps")
+        elif isinstance(parsed, list):
+            steps = parsed
+        else:
+            return []
+    if not isinstance(steps, list) or not all(
+        isinstance(s, dict) and isinstance(s.get("kind"), str) for s in steps
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Fixed, and somewhat embarrassingly: the first version of this very validation reproduced the unhashable-kind bug it was meant to catch, caught by the regression test on the first run. The loader now requires kind to be a string in the documented set, so a misspelled `aciton` fails closed instead of scoring 1.0.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 20]
```diff
@@ -0,0 +1,308 @@
...
+    # data quality: actions with missing or non-boolean ok -> low (fail-closed)
+    unannotated = [s for s in actions if not isinstance(s.get("ok"), bool)]
+    quality = "low" if unannotated else "ok"
+
+    T = THRESHOLDS
+    out: list[dict[str, Any]] = []
+
+    # 1 Anticipation: a plan/intent step appears before the first tool action
+    first_act = next(
+        (i for i, s in enumerate(steps) if s.get("kind") == "action" and s.get("tool")),
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Fixed. Anticipation now keys on the first action of any kind, matching the fallback-identity treatment in rule 3. A tool-less action starting before the plan no longer slips through.

File: README.md [hunk 21]
```diff
@@ -1,101 +1,91 @@
...
-Requires [Node.js](https://nodejs.org/en/download) `>=22.22.0` for npm and npx usage. Node.js 24 LTS
-is recommended; see the [runtime support guide](https://www.promptfoo.dev/docs/installation/#nodejs-runtime-support).
-
-```sh
-npm install -g promptfoo
-promptfoo init --example getting-started
-```
-
-Also available via `brew install promptfoo` and `pip install promptfoo`. You can also use `npx promptfoo@latest` to run any command without installing.
+# eval-agent-behavior-review (Agent Behavior Presentation Grader)
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply README.md]: My apologies on this one — the root cause was mine: a batch upload got targeted at the repository root instead of the example folder. The root README is restored to the upstream version, the stray root-level copies are removed, and the PR is back to five added files under the example directory. Sorry for the noise this added to the review.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+        for i in range(len(steps))
+    )
+    p2p = (not (len(steps) > T["long_session"])) or mid_verify
+    out.append(
+        {
+            "pass": p2p,
+            "score": 1.0 if p2p else 0.0,
+            "reason": "mid-session checkpoint present"
+            if p2p
+            else "long run without intermediate verification",
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Fixed. Rule 4 now reports "short session, checkpoint not required" when the exemption is what made it pass, and rule 9 got the same honest treatment on the feedback side.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 23]
```diff
@@ -0,0 +1,308 @@
...
+
+    # 1 Anticipation: a plan/intent step appears before the first tool action
+    first_act = next(
+        (i for i, s in enumerate(steps) if s.get("kind") == "action" and s.get("tool")),
+        None,
+    )
+    if first_act is None:
+        ante = {"pass": True, "score": 1.0, "reason": "no tool actions to anticipate"}
+    else:
+        declared = any(steps[j].get("kind") in _PLAN_KINDS for j in range(first_act))
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Good catch. Anticipation, Staging and Slow In/Out now require non-blank text on plan/intent steps, so a whitespace-only declaration can no longer carry a trace to a perfect score. Regression test added.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 24]
```diff
@@ -0,0 +1,308 @@
...
+            return []
+        if isinstance(parsed, dict):
+            steps = parsed.get("steps")
+        elif isinstance(parsed, list):
+            steps = parsed
+        else:
+            return []
+    if not isinstance(steps, list) or not all(
+        isinstance(s, dict) and isinstance(s.get("kind"), str) for s in steps
+    ):
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py:77]: Same mis-targeted upload as the README issue above, and again my mistake. The validation and its regression tests are now in the example-local files that promptfooconfig.yaml actually loads (see 6f4c791).

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 25]
```diff
@@ -0,0 +1,308 @@
...
+            "score": 0.0,
+            "reason": "no session trace in vars.session",
+            "componentResults": [],
+        }
+    score = sum(c["score"] for c in comps) / len(comps)
+    named = {("p" + str(i + 1)): c["score"] for i, c in enumerate(comps)}
+    fails = [c["reason"] for c in comps if not c["pass"]]
+    data_note = " [data_quality=low]" if res["quality"] == "low" else ""
+    # fail-closed: low data quality hard-fails the grade even above the score gate
+    passed = score >= THRESHOLDS["pass_score"] and res["quality"] != "low"
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py:341]: Fixed in 3d1315d. Traces without any action, verify, or report step are now rejected before the score gate with an explicit "no evidence of work" result, instead of collecting free passes from inapplicable rules. Regression test added.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 26]
```diff
@@ -0,0 +1,308 @@
...
+            "pass": p2p,
+            "score": 1.0 if p2p else 0.0,
+            "reason": "mid-session checkpoint present"
+            if p2p
+            else "long run without intermediate verification",
+        }
+    )
+
+    # 5 FollowThrough: session ends with verify/report
+    ft = bool(kinds) and kinds[-1] in _VERIFY_KINDS
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Fixed in 3d1315d. Follow Through and the verify-out half of Slow In/Out now require non-blank string content on the closing step, so a whitespace-only report no longer credits the conclusion phase.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 27]
```diff
@@ -0,0 +1,328 @@
...
+    "max_repeated_terms": 5,  # rule 11 heuristic
+}
+
+_PLAN_KINDS = {"plan", "intent"}
+_VERIFY_KINDS = {"verify", "report"}
+_VALID_KINDS = _PLAN_KINDS | _VERIFY_KINDS | {"action", "message"}
+
+
+def _has_content(s: dict[str, Any]) -> bool:
+    return bool(str(s.get("text", "")).strip())
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py]: Fixed in 3d1315d. _has_content now checks isinstance(text, str) before the blank test, so null and [] are no longer stringified into fake content. The regression suite covers both malformed values.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 28]
```diff
@@ -0,0 +1,328 @@
...
+            "reason": f"max same-error retries={worst} <= {T['max_retries']}"
+            if ok_
+            else f"retry loop x{worst}",
+        }
+    out.append(squash)
+
+    # 4 PoseToPose: checkpoint between the first and last action (preflight does not count)
+    mid_verify = bool(action_idx) and any(
+        action_idx[0] < i < action_idx[-1] and kinds[i] in _VERIFY_KINDS
+        for i in range(len(steps))
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py:165]: Fixed in 3d1315d. Both mid_verify and mid_msgs now require the qualifying step to carry non-blank string content, so an empty verify between actions counts as neither a checkpoint nor feedback. Test included.

File: examples/eval-agent-behavior-review/behavior_review.py [hunk 29]
```diff
@@ -0,0 +1,349 @@
...
+            "pass": False,
+            "score": 0.0,
+            "reason": "data_quality low: actions missing boolean ok annotation; retry loop undeterminable",
+        }
+    else:
+        err_streak: dict[str, int] = {}
+        worst = 0
+        for s in actions:
+            tool = s.get("tool") or "action"  # fallback identity when tool is omitted
+            key = str(tool) + "|" + str(s.get("text", ""))  # full error identity
```
[the remark this answers is not in the document]
Comment (ZhaoAndy821) [reply examples/eval-agent-behavior-review/behavior_review.py:144]: The pipe separator can in principle collide across field boundaries, but only for tool names containing a literal "|", and the failure direction is over-flagging rather than a bypass. Leaving the encoding as is; happy to switch to tuple keys if a maintainer prefers.

**Comment by ZhaoAndy821** [comment]
A word for the maintainers, since the review bot has now been through this example ten times.

Rounds 3 through 9 each turned up something real: fail-closed gaps, input shapes that could fake their way past a rule, component reasons that described evidence that wasn't there. All of it got fixed against a concrete failing input, and each fix carries a regression test now (49 in total).

What's left in the recent rounds reads more synthetic, honestly. Pipe characters inside tool names colliding with the key separator, boolean topic ids masquerading as numeric ones, that kind of thing. Real exporters are unlikely to produce these, and where they'd bite, the grader errs toward flagging rather than passing.

So I'd like to freeze the code here. The thresholds and heuristics are documented as initial estimates pending calibration, and I'm glad to take any further hardening in whatever direction a human reviewer thinks is worth going.
